etna/transforms/encoders/mean_encoder.py

```diff
@@ -0,0 +1,269 @@
+                    feature = (temp["cumsum"] - y + self.global_means[segment] * self.smoothing) / (
+                        temp["cumcount"] + self.smoothing
+                    )
+                    intersected_df.loc[segment_df.index, self.out_column] = feature
+                    if self.handle_missing is MissingMode.global_mean:
+                        nan_index = segment_df[segment_df[self.in_column].isnull()].index
+                        expanding_mean = y.expanding().mean().shift().fillna(0)
+                        intersected_df.loc[nan_index, self.out_column] = expanding_mean.loc[nan_index]
+            else:
+                temp = pd.DataFrame(index=intersected_df.index, columns=["cumsum", "cumcount"], dtype=float)
+
+                timestamps = intersected_df["timestamp"].unique()
+                categories = intersected_df[self.in_column].unique()
+                cumstats = pd.DataFrame(data={"sum": 0, "count": 0, self.in_column: categories})
+                for timestamp in timestamps:
+                    timestamp_df = intersected_df[intersected_df["timestamp"] == timestamp]
+                    timestamp_index = timestamp_df.index
+                    cumsum_dict = dict(cumstats[[self.in_column, "sum"]].values)
+                    cumcount_dict = dict(cumstats[[self.in_column, "count"]].values)
+                    temp.loc[timestamp_index, "cumsum"] = intersected_df.loc[timestamp_index, self.in_column].map(
+                        cumsum_dict
+                    )
+                    temp.loc[timestamp_index, "cumcount"] = intersected_df.loc[timestamp_index, self.in_column].map(
+                        cumcount_dict
+                    )
+
+                    stats = (
+                        timestamp_df["target"]
+                        .groupby(timestamp_df[self.in_column], dropna=False)
+                        .agg(["count", "sum"])
+                        .reset_index()
+                    )
+                    cumstats = pd.concat([cumstats, stats]).groupby(self.in_column, as_index=False, dropna=False).sum()
+
+                feature = (temp["cumsum"] + self.global_means * self.smoothing) / (temp["cumcount"] + self.smoothing)
+                intersected_df[self.out_column] = feature
+
+                if self.handle_missing is MissingMode.global_mean:
+                    y = intersected_df["target"]
+                    nan_index = intersected_df[intersected_df[self.in_column].isnull()].index
+                    timestamp_count = y.groupby(intersected_df["timestamp"]).transform("count")
+                    timestamp_sum = y.groupby(intersected_df["timestamp"]).transform("sum")
+                    expanding_mean = (
+                        timestamp_sum.iloc[::n_segments].cumsum() / timestamp_count.iloc[::n_segments].cumsum()
+                    )
+                    expanding_mean = expanding_mean.repeat(n_segments)
```

Comment:
Do we really need this `repeat`? Why do we need to set index and make shift in the next line?